app/src/main/java/org/eyeseetea/malariacare/data/database/iomodules/dhis/exporter/PushController.java

```diff
@@ -22,167 +22,109 @@
-            //Asks sdk to push localdata
-            postProgress(context.getString(R.string.progress_push_posting_survey));
-            Log.d(TAG, "Pushing survey data to server...");
-            SdkPushController.sendEventChanges();
-        }catch (Exception ex) {
-            Log.e(TAG, "push: " + ex.getLocalizedMessage());
-            postException(ex);
-            isPushing =false;
-            return false;
+            if (EventExtended.getAllEvents().size() == 0){
+                callback.onError(new ConversionException());
+            }
+            else{
+                pushData(callback);
+            }
         }
-        return true;
     }
 
-    /**
-     * Launches visitor that turns an APP survey into a SDK event
-     */
-    private void convertToSDK(List<Survey> surveys)throws  Exception{
-        Log.d(TAG,"Converting APP survey into a SDK event");
-        converter =new ConvertToSDKVisitor(context);
-        for(Survey survey:surveys){
-            survey.accept(converter);
-        }
+    @Override
+    public boolean isPushInProgress() {
+        return PreferencesState.getInstance().isPushInProgress();
     }
 
-    /**
-     * Notifies a progress into the bus (the caller activity will be listening)
-     * @param msg
-     */
-    private void postProgress(String msg){
-        SdkController.postProgress(msg);
+    @Override
+    public void changePushInProgress(boolean inProgress) {
+        PreferencesState.getInstance().setPushInProgress(inProgress);
     }
 
-    /**
-     * Notifies an exception while pulling
-     * @param ex
-     */
-    private void postException(Exception ex){
-        AlarmPushReceiver.isDoneFail();
-        ex.printStackTrace();
-        SdkController.postException(ex);
+    private void pushData(final IPushControllerCallback callback) {
+        mPushDhisSDKDataSource.pushData(
+                new IDataSourceCallback<Map<String, ImportSummary>>() {
+                    @Override
+                    public void onSuccess(
+                            Map<String, ImportSummary> mapEventsImportSummary) {
+                        mConvertToSDKVisitor.saveSurveyStatus(mapEventsImportSummary);
+                        callback.onComplete();
+                    }
+
+                    @Override
+                    public void onError(Throwable throwable) {
+                        setSurveysAsQuarantine();
+                        callback.onError(throwable);
+                    }
+                });
     }
 
-    /**
-     * Notifies that the push is over
-     */
-    public static void postFinish(boolean success){
-        try {
-            if(success){
-                AlarmPushReceiver.isDoneSuccess();
-            }else{
-                AlarmPushReceiver.isDoneFail();
+    private List<Survey> findSurveysToPush() {
+        List<Survey> surveys;
+        surveys = Survey.getAllUnsentUnplannedSurveys();
+        for (int i = 0; i < surveys.size(); i++) {
```

Comment:
@xurxodev  this loop is not necessary. 
The findSurveysToPush method should be replaced by Survey.getAllUnsentUnplannedSurveys();
Or you can remove the progressactivity logic from findSurveysToPush
thanks!